cumulus_library_data_metrics/q_date_in_lifetime/patient_lifespan.sql

```diff
@@ -0,0 +1,33 @@
+-- partial_date: iso date string
+-- suffix: to append to year if date is incomplete. Ex : '-01-01T23:59:59'
+CREATE MACRO DATE_WITH_FALLBACK_SUFFIX(partial_date, suffix) AS (
+    CASE -- Any date of format YYYY-MM-DD or more detailed, just parse
+        WHEN LENGTH(partial_date) > 9 THEN from_iso8601_timestamp(partial_date)
+        -- If there is a parsable year, capture that year with the suffix appended
+        WHEN regexp_matches(partial_date, '^[0-9]{4}', 'c') THEN from_iso8601_timestamp(concat(regexp_extract(partial_date, '^[0-9]{4}'), suffix))
+        -- Bail, take parser's best attempt, assume null
+        ELSE from_iso8601_timestamp(partial_date)
+    END
+);
+
+CREATE MACRO DATE_WITHIN_LIFESPAN(observation_date, birth, death) AS (
+    CASE
+        -- when the patient has a date of death, verify two things:
+        -- 1. latest possible interpretation of obs.date is after birth
+        -- 2. earliest possible interpretation of obs.date is before death
+        WHEN P.patient_end IS NOT NULL THEN (
+            DATE_WITH_FALLBACK_SUFFIX(observation_date, '-12-31T23:59:59') > birth AND
+            DATE_WITH_FALLBACK_SUFFIX(observation_date, '-01-01T00:00:00') < death
+        )
+        -- when the patient just has a birth date, verify just:
+        -- 1. latest possible interpretation of obs.date is after birth
+        ELSE DATE_WITH_FALLBACK_SUFFIX(observation_date, '-12-31T23:59:59') > birth
```

Comment:
Birth date can be null too. So I think the logic here might be simpler if we broke it up into two separate checks? Something like this pseudo code:

```sql
(death IS NULL OR date < death)
AND
(birth IS NULL OR date > birth)
```